caravel/assets/javascripts/SqlLab/components/SqlEditorLeftBar.jsx

```diff
@@ -40,41 +41,49 @@ class SqlEditorLeftBar extends React.PureComponent {
     if (!(db)) {
       this.setState({ tableOptions: [] });
     } else {
-      this.fetchTables(val, this.props.queryEditor.schema);
       this.fetchSchemas(val);
+      this.fetchTables(val, this.props.queryEditor.schema);
     }
   }
   resetState() {
     this.props.actions.resetState();
   }
-  fetchTables(dbId, schema) {
-    const actualDbId = dbId || this.props.queryEditor.dbId;
-    if (actualDbId) {
-      const actualSchema = schema || this.props.queryEditor.schema;
-      this.setState({ tableLoading: true });
+  getTableNamesBySubStr(input, callback) {
+    this.fetchTables(
+        this.props.queryEditor.dbId,
+        this.props.queryEditor.schema,
+        input);
+    callback(null, { options: this.state.tableOptions });
+  }
+  fetchTables(dbId, schema, substr) {
+    if (!dbId) {
       this.setState({ tableOptions: [] });
-      const url = `/caravel/tables/${actualDbId}/${actualSchema}`;
-      $.get(url, (data) => {
-        let tableOptions = data.tables.map((s) => ({ value: s, label: s }));
-        const views = data.views.map((s) => ({ value: s, label: '[view] ' + s }));
-        tableOptions = [...tableOptions, ...views];
-        this.setState({ tableOptions });
-        this.setState({ tableLoading: false });
-      });
+      return;
     }
+    this.setState({ tableLoading: true });
+    this.setState({ tableOptions: [] });
+    const url = `/caravel/tables/${dbId}/${schema}?substr=${substr}`;
+    $.get(url, (data) => {
+      let tableOptions = data.tables.map((s) => ({ value: s, label: s }));
+      const views = data.views.map((s) => ({ value: s, label: '[view] ' + s }));
+      tableOptions = [...tableOptions, ...views];
+      this.setState({ tableOptions });
+      const tableLength = data.views_length + data.tables_length;
+      this.setState({ tableLength });
+      this.setState({ tableLoading: false });
+    });
   }
   changeSchema(schemaOpt) {
     const schema = (schemaOpt) ? schemaOpt.value : null;
     this.props.actions.queryEditorSetSchema(this.props.queryEditor, schema);
     this.fetchTables(this.props.queryEditor.dbId, schema);
   }
   fetchSchemas(dbId) {
-    const actualDbId = dbId || this.props.queryEditor.dbId;
-    if (actualDbId) {
+    if (dbId) {
       this.setState({ schemaLoading: true });
-      const url = `/databasetablesasync/api/read?_flt_0_id=${actualDbId}`;
+      const url = `/caravel/schemas/${dbId}`;
```

Comment:
same here, we should be handling the fetching of data in actions and setting state there.